inst/include/dust/filter.hpp

```diff
@@ -37,9 +58,18 @@ filter(T * obj,
     state.snapshots.resize(obj->n_state_full(), n_particles, step_snapshot);
   }
 
-  for (auto & d : obj->data()) {
-    obj->run(d.first);
-    obj->compare_data(weights, d.second);
+  auto d = obj->data().cbegin();
+  const auto d_end = obj->data().cend();
+
+  while (d->first <= obj->step() && d != d_end) {
+    d++;
+    state.trajectories.advance();
+  }
+
+  for (; d != d_end; ++d) {
```

Comment:
I've never seen a loop like this before, not to say it's wrong.
Would it not be more typical to have
```
for (auto it = d; it != d_end; ++it)
```

Reply:
this is fairly conventional, c++ is silly